### ISSUE TITLE
API for clipping gradients by L2 regularization ?

### ISSUE THREAD

**Issue by zhang-wen** [clab/dynet#179]
i want to normalize the L2-norm of the gradient of the cost function each time to be at most a predefined threshold of 1, when the norm was larger than the threshold

is there clip_gradients (L2 regularization) API for python ?

**Comment by pmichel31415** [comment]
sgd = new SimpleSGDTrainer(&model);
    sgd->clip_threshold = 1;

Is that what you're looking for?

**Comment by zhang-wen** [comment]
yes, exactly, but no API in python.
is it easy to implement the API in python ?

**Comment by pmichel31415** [comment]
Shouldn't be too hard, I'll look into it

**Comment by pmichel31415** [comment]
Now that I think of it there is a flag to enable/disable clipping altogether in c++, it would eb a good idea to access it from python.

One way of doing this would be to disable clipping whenever a negative value is passed to `set_clip_threshold` instead of creating new accessors/setters.

What would be better @yoavg ?

**Comment by yoavg** [comment]
does it make sense for set_clipping_threshold(0) to disable clipping?

**Comment by pmichel31415** [comment]
I think it makes sense if this is specified in the docs.

**Comment by zhang-wen** [comment]
it works, thank you !
